### ISSUE TITLE
Visual Studio 17.1 update has broken assignment

### ISSUE THREAD

**Issue by lukester1975** [codeinred/tuplet#21]
Seems like there's a few issues with VS 17.1 and fold expressions (well, maybe that's where the issue is - not sure really) and tuplet suffers (assignment tests fail).

Whilst trying to make a minimal standalone example to post to MS, I found it was allowing invalid code: https://developercommunity.visualstudio.com/t/fold-expressions-unreliable-in-171-with-c20/1676476 (as I write, that ticket isn't live yet; stuck in some queue due to all the bugs in the new compiler being reviewed, maybe ... but hopefully will be soon).

I'm guessing once they fix that similar issue, the tuplet assignment will work again. That said, I have a workaround:

https://github.com/lukester1975/tuplet/blob/f18bc731f062524376d993810f3fd673bc02185e/include/tuplet/tuple.hpp#L294-L303

which allows assignment tests to pass again. Branch is here: https://github.com/lukester1975/tuplet/tree/cl17.1-workaround

1. There's a few versions: simplest that I realised after the others was just to remove the `identity<B2>` and simply use `B2`. I can't see what it's there for, but might be missing something...
2. If so, the third version `void(static_cast<B1&>(*this).value = std::forward<identity_t<B2>>(u).value), ...);` keeps it.
3. Note that in this version, only the right-side of the assignment is actually required. I only changed the lhs to a cast to be consistent.
4. There's an additional test for using `tie` with `move` as my initial "fix" (none of these versions, though) inadvertently missed a move and this new test caught it.

As an aside, could I ask the reason for the void in `(void(B1::value = static_cast<U&&>(u).identity_t<B2>::value), ...);` ? I've seen similar in boost. Does it avoid a warning about unused result on some compilers ... or something else?

Happy to make a PR if you want one of the workarounds.

Thanks!

**Comment by lukester1975** [comment]
Ticket still not visible. Weird. Anyway here's the compiler explorer link with a (presumably) related issue I posted to MS: https://godbolt.org/z/E4WhcT6e3

**Comment by codeinred** [comment]
Hi! I integrated this fix from your branch, thank you so much! Your commit is included:

![image](https://user-images.githubusercontent.com/36810712/155602816-a3b7f907-3f3f-4d05-833e-d4d894a0b107.png)

I did it directly rather than via a pull request because tests were failing without the fix provided for google benchmark that I committed in #20, and it would have been something of a mess otherwise (since both pull requests would have required the other to pass all tests)

The reason for identity_t is that in some contexts some compilers wouldn't properly recognize the base classes as base classes, in certain contexts, and identity_t fixed that. [A longer explanation is provided here.](https://github.com/codeinred/tuplet/issues/10#issuecomment-939915607) It appears that B2 is properly recognized without `identity_t` in the context you provided, so the first fix you suggested was good!

The `void` appears because sometimes people overload the comma operator. The comma operator *can't* be overloaded for types like `void`, though, so this ensures that no comma operator overloads are involved (and it functions like the comma operator is supposed to).
